dart-mode.el

```diff
@@ -562,6 +582,298 @@ This uses `dart-format-path' to find the formatter."
+          (format-time-string iso-format-string (current-time))))
+    (with-current-buffer log-buffer
+      (goto-char (point-max))
+      (insert "\n\n\n")
+      (insert (concat timestamp-and-log-string
+                      "\n"
+                      msg))
+      (insert "\n"))))
+
+(defun dart--start-analysis-server-for-current-buffer ()
+  "Initialize Dart analysis server for current buffer.
+
+This starts Dart analysis server and adds either the pub root
+directory or the current file directory to the analysis roots."
+  (if (not dart--analysis-server) (dart-start-analysis-server))
+  ;; TODO(hterkelsen): Add this file to the priority files.
+  (dart-add-analysis-root-for-file)
+  (add-hook 'first-change-hook 'dart-add-analysis-overlay t t)
+  (add-hook 'after-change-functions 'dart-change-analysis-overlay t t)
+  (add-hook 'after-save-hook 'dart-remove-analysis-overlay t t)
+  (add-to-list 'flycheck-checkers 'dart-analysis-server))
+
+(defun dart-start-analysis-server ()
+  "Start the Dart analysis server."
+  (when dart--analysis-server
+    (process-kill-without-query
+     (dart--analysis-server-process dart--analysis-server))
+    (kill-buffer (dart--analysis-server-buffer dart--analysis-server)))
+  (let ((dart-process
+         (start-process "dart-analysis-server"
+                        "*dart-analysis-server*"
+                        dart-executable-path
+                        dart-analysis-server-snapshot-path
+                        "--no-error-notification")))
+    (setq dart--analysis-server
+          (dart--analysis-server-create dart-process))))
+
+(defun dart--analysis-server-create (process)
+  "Create a Dart analysis server from PROCESS."
+  (lexical-let* ((buffer (generate-new-buffer (process-name process)))
+                 (instance (dart--make-analysis-server
+                            :process process
+                            :buffer buffer)))
+    (buffer-disable-undo (dart--analysis-server-buffer instance))
+    (set-process-filter
+     process
+     (lambda (proc string)
+       (dart--analysis-server-process-filter instance string)))
+    instance))
+
+(defun dart-add-analysis-overlay ()
+  "Report to Dart analysis server that it should overlay this buffer.
+
+The Dart analysis server allows clients to 'overlay' file contents with
+a client-supplied string.  This is needed because we want Emacs to report
+errors for the current contents of the buffer, not whatever is saved to disk."
+  (dart--analysis-server-send
+   "analysis.updateContent"
+   `((files .
+            ((,buffer-file-name . ((type . "add")
+                                   (content . ,(buffer-string)))))))))
+
+(defun dart-change-analysis-overlay
+    (change-begin change-end change-before-length)
+  "Report to analysis server that it should change the overlay for this buffer.
+
+The region that changed ranges from CHANGE-BEGIN to CHANGE-END, and the
+length of the text before the change is CHANGE-BEFORE-LENGTH."
```

Comment:
Maybe add "See also \`dart-add-analysis-overlay'." to the docstrings for this and `dart-remove-analysis-overlay`.